metadata/src/test/java/org/apache/kafka/image/ClusterImageTest.java

```diff
@@ -205,4 +207,22 @@ private static List<ApiMessageAndVersion> getImageRecords(ClusterImage image) {
         image.write(writer, new ImageWriterOptions.Builder().build());
         return writer.records();
     }
+
+    @Test
+    public void testHandleLossOfControllerRegistrations() {
+        ClusterImage testImage = new ClusterImage(Collections.emptyMap(),
+            Collections.singletonMap(1000, new ControllerRegistration.Builder().
+                setId(1000).
+                setIncarnationId(Uuid.fromString("9ABu6HEgRuS-hjHLgC4cHw")).
+                setListeners(Collections.singletonMap("PLAINTEXT",
+                    new Endpoint("PLAINTEXT", SecurityProtocol.PLAINTEXT, "localhost", 19092))).
+                setSupportedFeatures(Collections.emptyMap()).build()));
+        RecordListWriter writer = new RecordListWriter();
+        final AtomicReference<String> lossString = new AtomicReference<>("");
+        testImage.write(writer, new ImageWriterOptions.Builder().
+            setMetadataVersion(MetadataVersion.IBP_3_6_IV2).
+            setLossHandler(loss -> lossString.compareAndSet("controller registration data", loss.loss())).
+                build());
+        assertEquals("", lossString.get());
```

Comment:
Does this pass? Looking at `MetadataVersion`, controller registration support was added in `3.7-IV0`. I would expect that since this test sets the MV to `3.6-IV2`, this test would call the loss handler.

Can we also add a test where the controller registration record is included to the snapshot?

Reply:
Sorry, I forgot to push the latest version. Should be fixed now.